core/src/main/java/org/apache/iceberg/BaseFileScanTask.java

```diff
@@ -45,31 +50,67 @@ protected FileScanTask self() {
 
   @Override
   protected FileScanTask newSplitTask(FileScanTask parentTask, long offset, long length) {
-    return new SplitScanTask(offset, length, parentTask);
+    return new SplitScanTask(offset, length, deletesSizeBytes(), parentTask);
   }
 
   @Override
   public List<DeleteFile> deletes() {
-    return ImmutableList.copyOf(deletes);
+    if (deletesAsList == null) {
+      this.deletesAsList = Collections.unmodifiableList(Arrays.asList(deletes));
+    }
+
+    return deletesAsList;
+  }
+
+  @Override
+  public long sizeBytes() {
+    return length() + deletesSizeBytes();
+  }
+
+  @Override
+  public int filesCount() {
+    return 1 + deletes.length;
```

Comment:
Override this to avoid materializing `deletesAsList` to simply compute the files count.

Reply:
It would probably consider that separately from this PR. It would be unfortunate to materialize the list but it would not be the end of the world.

Reply:
Why not initialize the deletes list in the constructor?
Then no need to have these overrides. 

That would mean BaseFileScanTask constructor needs to make a copy, but that's probably actually a good thing, since it would make the class more immutable.
It will have to make the copy anyway, since `List deletes()` is the only way to get the deletes information from the class, so it will be called sooner or later.